lib-gphotos/client.go

```diff
@@ -91,39 +156,44 @@ func (client *Client) UploadFile(filePath string, pAlbumID ...string) (*photosli
 		albumID = pAlbumID[0]
 	}
 
-	filename := path.Base(filePath)
-	log.Printf("Uploading %s", filename)
+	log.Printf("Uploading %s", filePath)
 
 	file, err := os.Open(filePath)
 	if err != nil {
-		return nil, stacktrace.Propagate(err, "failed opening file")
+		return nil, stacktrace.Propagate(err, "failed opening file %s", filePath)
 	}
 	defer file.Close()
 
-	uploadToken, err := client.GetUploadToken(file, filename)
+	fileStat, err := file.Stat()
+	if err != nil {
+		return nil, stacktrace.Propagate(err, "failed getting file size %s", filePath)
+	}
+	fileSize := fileStat.Size()
+
+	uploadToken, err := c.GetUploadToken(file, filePath, fileSize, uploadURL)
 	if err != nil {
-		return nil, stacktrace.Propagate(err, "failed getting uploadToken for %s", filename)
+		return nil, stacktrace.Propagate(err, "failed getting uploadToken for %s", filePath)
 	}
 
 	retry := true
 	retryCount := 0
 	for retry != false {
 		retry = false
-		batchResponse, err := client.MediaItems.BatchCreate(&photoslibrary.BatchCreateMediaItemsRequest{
+		batchResponse, err := c.MediaItems.BatchCreate(&photoslibrary.BatchCreateMediaItemsRequest{
 			AlbumId: albumID,
 			NewMediaItems: []*photoslibrary.NewMediaItem{
 				{
-					Description:     filename,
+					Description:     filePath,
 					SimpleMediaItem: &photoslibrary.SimpleMediaItem{UploadToken: uploadToken},
 				},
 			},
 		}).Do()
 		if err != nil {
 			// handle rate limit error by sleeping and retrying
-			if err.(*googleapi.Error).Code == 429 {
-				after, err := strconv.ParseInt(err.(*googleapi.Error).Header.Get("Retry-After"), 10, 64)
+			if gerr, ok := err.(*googleapi.Error); ok && gerr.Code == 429 {
```

Comment:
Following the [Google dev guide](https://developers.google.com/photos/library/guides/best-practices#retrying-failed-requests) we should work with exponential backoffs for codes 5xx and at lest 30s for code 429. 

This is a TODO  to be implemented. Just mentioning!